File: plugins/actian-design-system/scripts/preview-server.py
#!/usr/bin/env python3
"""Preview server with annotation POST endpoint.

Serves static files from the working directory AND accepts POST to /_annotations
to write annotation JSON to .annotations.json in the served directory.

Usage: python3 preview-server.py [port] [directory]
"""

import json
import os
import sys
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path


class PreviewHandler(SimpleHTTPRequestHandler):
    """Static file server + POST /_annotations + plugin asset serving."""

    # Set by main() — path to the plugin's templates/ directory
    plugin_templates_dir = None

    def do_GET(self):
        if self.path.startswith('/_version'):
            self._handle_version()
        elif self.path.startswith('/_plugin/'):
            self._serve_plugin_asset()
        else:
            super().do_GET()

    def _serve_plugin_asset(self):
        """Serve files from the plugin's templates directory at /_plugin/..."""
        if not self.plugin_templates_dir:
            self.send_error(404, 'Plugin templates path not configured')
            return
        rel_path = self.path[len('/_plugin/'):]
        file_path = Path(self.plugin_templates_dir) / rel_path
        # Security: prevent path traversal
        try:
            file_path = file_path.resolve()
            templates_dir = Path(self.plugin_templates_dir).resolve()
            if not str(file_path).startswith(str(templates_dir)):
                self.send_error(403, 'Forbidden')
                return
        except Exception:
            self.send_error(400, 'Bad path')
            return
        if not file_path.is_file():
            self.send_error(404, 'Not found')
            return
        # Determine content type
        ext = file_path.suffix.lower()
        content_types = {
            '.html': 'text/html', '.css': 'text/css', '.js': 'application/javascript',
            '.json': 'application/json', '.svg': 'image/svg+xml',
        }
        content_type = content_types.get(ext, 'application/octet-stream')
        try:
            data = file_path.read_bytes()
            self.send_response(200)
            self.send_header('Content-Type', content_type)
            self.send_header('Content-Length', len(data))
            self.send_header('Cache-Control', 'no-cache')
            self.end_headers()
            self.wfile.write(data)
        except Exception as e:
            self.send_error(500, str(e))

    def do_POST(self):
        if self.path == '/_annotations':
            self._handle_annotations()
        else:
            self.send_error(404, 'Not Found')

    def _handle_version(self):
        """Return mtime of a file for live-reload polling."""
        from urllib.parse import urlparse, parse_qs
        query = parse_qs(urlparse(self.path).query)
        file_path = query.get('file', [''])[0]
        if not file_path:
            self.send_error(400, 'Missing file param')
            return
        full_path = (Path(self.directory) / file_path.lstrip('/')).resolve()
        # Security: prevent path traversal
        if not str(full_path).startswith(str(Path(self.directory).resolve())):
            self.send_error(403, 'Forbidden')
            return
        try:
            mtime = full_path.stat().st_mtime
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Cache-Control', 'no-cache')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'mtime': mtime}).encode())
        except FileNotFoundError:
            self.send_error(404, 'File not found')

    def _handle_annotations(self):
        try:
            MAX_BODY = 10 * 1024 * 1024  # 10MB
            length = int(self.headers.get('Content-Length', 0))
            if length > MAX_BODY:
                self.send_error(413, 'Request body too large')
                return
            body = self.rfile.read(length)
            data = json.loads(body)

            # Validate minimal structure
            if 'annotations' not in data or not isinstance(data['annotations'], list):
                self.send_error(400, 'Missing or invalid annotations array')
                return

            # Write to .annotations.json in the served directory
            out_path = Path(self.directory) / '.annotations.json'
            with open(out_path, 'w') as f:
                json.dump(data, f, indent=2)

            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({
                'ok': True,
                'file': str(out_path),
                'count': len(data['annotations'])
            }).encode())

        except json.JSONDecodeError:
            self.send_error(400, 'Invalid JSON')
        except Exception as e:
            self.send_error(500, str(e))

    def do_OPTIONS(self):
        """Handle CORS preflight for POST."""
        self.send_response(204)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def log_message(self, format, *args):
        """Suppress request logging."""
        pass


def main():
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8765
    directory = sys.argv[2] if len(sys.argv) > 2 else '.'

    os.chdir(directory)
    PreviewHandler.directory = os.getcwd()

    # Plugin templates directory (sibling of scripts/)
    script_dir = Path(__file__).resolve().parent
    PreviewHandler.plugin_templates_dir = str(script_dir.parent / 'templates')

    server = HTTPServer(('', port), PreviewHandler)
    server.serve_forever()


if __name__ == '__main__':
    main()

File: plugins/actian-design-system/scripts/renderers/preview-server.js
#!/usr/bin/env node
/**
 * Preview server with annotation POST endpoint.
 *
 * Serves static files from the working directory AND accepts POST to /_annotations
 * to write annotation JSON to .annotations.json in the served directory.
 *
 * Usage: node preview-server.js [port] [directory]
 */

const http = require("http");
const fs = require("fs");
const path = require("path");
const url = require("url");

const port = parseInt(process.argv[2], 10) || 8765;
const directory = process.argv[3] || ".";
const dirAbs = path.resolve(directory);
const templatesDir = path.resolve(path.join(__dirname, "../..", "templates"));

const MIME_TYPES = {
  ".html": "text/html",
  ".css": "text/css",
  ".js": "application/javascript",
  ".json": "application/json",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".ico": "image/x-icon",
  ".woff": "font/woff",
  ".woff2": "font/woff2",
};

const MAX_BODY = 10 * 1024 * 1024; // 10MB

function sendJSON(res, status, data, extraHeaders) {
  const body = JSON.stringify(data);
  res.writeHead(
    status,
    Object.assign(
      {
        "Content-Type": "application/json",
        "Cache-Control": "no-cache",
        "Access-Control-Allow-Origin": "*",
      },
      extraHeaders || {},
    ),
  );
  res.end(body);
}

function sendError(res, status, message) {
  res.writeHead(status, { "Content-Type": "text/plain" });
  res.end(message);
}

function serveFile(res, filePath, baseDir) {
  const resolved = path.resolve(filePath);
  if (!resolved.startsWith(baseDir + path.sep) && resolved !== baseDir) {
    return sendError(res, 403, "Forbidden");
  }
  fs.stat(resolved, (err, stats) => {
    if (err || !stats.isFile()) {
      return sendError(res, 404, "Not found");
    }
    const ext = path.extname(resolved).toLowerCase();
    const contentType = MIME_TYPES[ext] || "application/octet-stream";
    fs.readFile(resolved, (readErr, data) => {
      if (readErr) return sendError(res, 500, "Read error");
      res.writeHead(200, {
        "Content-Type": contentType,
        "Content-Length": data.length,
        "Cache-Control": "no-cache",
      });
      res.end(data);
    });
  });
}

const server = http.createServer((req, res) => {
  const parsed = url.parse(req.url, true);
  const pathname = decodeURIComponent(parsed.pathname);

  // CORS preflight
  if (req.method === "OPTIONS") {
    res.writeHead(204, {
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Methods": "POST, OPTIONS",
      "Access-Control-Allow-Headers": "Content-Type",
    });
    return res.end();
  }

  // /_version — file mtime for live-reload polling
  if (req.method === "GET" && pathname === "/_version") {
    const file = parsed.query.file;
    if (!file) return sendError(res, 400, "Missing file param");
    const fullPath = path.resolve(path.join(dirAbs, file.replace(/^\//, "")));
    if (!fullPath.startsWith(dirAbs + path.sep) && fullPath !== dirAbs) {
      return sendError(res, 403, "Forbidden");
    }
    fs.stat(fullPath, (err, stats) => {
      if (err) return sendError(res, 404, "File not found");
      sendJSON(res, 200, { mtime: Math.floor(stats.mtimeMs / 1000) });
    });
    return;
  }

  // POST /_annotations
  if (req.method === "POST" && pathname === "/_annotations") {
    const length = parseInt(req.headers["content-length"] || "0", 10);
    if (length > MAX_BODY) return sendError(res, 413, "Request body too large");
    const chunks = [];
    req.on("data", (chunk) => {
      chunks.push(chunk);
      if (Buffer.concat(chunks).length > MAX_BODY) {
        sendError(res, 413, "Request body too large");
        req.destroy();
      }
    });
    req.on("end", () => {
      try {
        const data = JSON.parse(Buffer.concat(chunks).toString());
        if (!data.annotations || !Array.isArray(data.annotations)) {
          return sendError(res, 400, "Missing or invalid annotations array");
        }
        const outPath = path.join(dirAbs, ".annotations.json");
        fs.writeFile(outPath, JSON.stringify(data, null, 2), (err) => {
          if (err) return sendError(res, 500, err.message);
          sendJSON(res, 200, {
            ok: true,
            file: outPath,
            count: data.annotations.length,
          });
        });
      } catch (e) {
        sendError(res, 400, "Invalid JSON");
      }
    });
    return;
  }

  // GET /_plugin/* — serve from templates/ directory
  if (req.method === "GET" && pathname.startsWith("/_plugin/")) {
    const relPath = pathname.slice("/_plugin/".length);
    const filePath = path.join(templatesDir, relPath);
    return serveFile(res, filePath, templatesDir);
  }

  // GET — static file serving from dirAbs
  if (req.method === "GET") {
    let filePath = path.join(dirAbs, pathname);
    // Serve index.html for directory requests
    if (pathname.endsWith("/")) filePath = path.join(filePath, "index.html");
    return serveFile(res, filePath, dirAbs);
  }

  sendError(res, 404, "Not Found");
});

server.listen(port, () => {
  process.stderr.write(
    `Preview server listening on http://localhost:${port} serving ${dirAbs}\n`,
  );
});
